qiskit/algorithms/optimizers/de.py

```diff
@@ -0,0 +1,134 @@
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Differential Evolution optimizer"""
+
+from typing import Callable, Tuple, List, Dict, Union
+
+import numpy as np
+from scipy.optimize import differential_evolution
+from .optimizer import Optimizer, OptimizerSupportLevel
+
+
+class DifferentialEvolution(Optimizer):
+    """
+    Differential Evolution optimizer
+
+    The Differential Evolution Optimizer is a global and derivative-free
+    optimizer that seeks to find the optimal value of an objective function,
+    f(x) (where x represents the vector of possible parameters) through
+    stochastically searching the optimization space of possible solutions.
+    It belongs to the family of evolutionary optimization techniques
+    such as genetic and particle swarm algorithms. It is best suited
+    for combinatorial optimization problems and operates on real-parameter
+    and real-valued objective functions.
+
+    Specifically, the algorithm begins with an initial population
+    of candidate solutions (can be random or user-specified) and new candidates
+    are iteratively created through a combination of mutation, recombination
+    and stochastic selection. The main parameters that govern the optimization
+    include the evolutionary strategy, population size, mutation constant
+    (differential weight) and recombination factor (crossover probability).
+    The algorithm terminates if either the maximum number of iterations/generations
+    have been reached or the standard deviation of the generation of candidates
+    is below a user-specified threshold. Specificially, the total number of
+    function evaluations equals (maxiter + 1)* popsize * len(x), without polishing
+    (see below).
+
+    Uses scipy.optimize.differential_evolution
+    For further detail, please refer to
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html
+    """
+
+    _OPTIONS = ['maxiter', 'strategy', 'popsize', 'mutation',
+                'recombination', 'init', 'tol', 'atol',
+                'polish', 'disp', 'workers']
+
+    def __init__(self,
+                 maxiter: int = 10,
+                 strategy: str = 'currenttobest1bin',
+                 popsize: int = 10,
+                 mutation: Union[float, tuple] = (0.5, 1),
+                 recombination: float = 0.7,
+                 init: str = 'latinhypercube',
+                 tol: float = 0.01,
+                 atol: float = 0,
+                 polish: bool = False,
+                 disp: bool = False,
+                 workers: Union[int, Callable] = 1) -> None:

```

Comment:
Might we want to expose the capability to supply a callback. Historically we did not do this but recently SPSA was updated it offers a callback at it level. It may allow access to information of interest to users.